### PR TITLE
[Aptos Framework][Account] added rotation capability offer and revoke functions

### PR DESCRIPTION
**Pull request by 0xchloe**
### Description
- added `offer_rotation_capability` function to allow an account owner to delegate their rotation capability to another account
- added `revoke_rotation_capability` function to allow an account owner to revoke their rotation capability offer to the `to_be_revoked_address`
- added struct `RotationCapabilityOfferProofChallengeV2` for backward compatibility and deprecated the previous `RotationCapabilityOfferProofChallenge`
- changed the more ambiguous `EACCOUNT_DOES_NOT_EXIST` error in capability-related functions to either `EOFFERER_ADDRESS_DOES_NOT_EXIST` or `ERECIPIENT_ADDRESS_DOES_NOT_EXIST`
- changed `ENO_SUCH_SIGNER_CAPABILITY` error to the more accurate naming `ENO_SUCH_SIGNER_CAPABILITY_OFFER`
- changed sequence number for `harness.new_account_with_key_pair()` from 10 to 0, so if other devs are trying to generate a valid signature for unit test purpose, they don't have to manually change the sequence number from 10 to 0. 

### Test Plan

unit & e2e tests

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4202)

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:308]: Add another field for the address of the source account here.

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:313]: Typo: SignerCapabilityOfferProofChallenge should be RotationCapabilityOfferProofChallenge

Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:313]: Technically, you don't "verify correctness of RotationCapabilityOfferProofChallenge" here. You've just built the correct one above, so I would rephrase this as: "verify the signature on RotationCapabilityOfferProofChallenge by the account owner"

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:333]: Why not use `std::option::swap_or_fill` here?

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:586]: Same issue as in the other PR: might be good to add instructions on how you generated this signature.

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:308]: cc @davidiw is this a breaking change?

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:313]: ok will update the comment! good point

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:333]: ahhhh didn't know that exists! I was only looking at [this move guide](https://move-language.github.io/move/standard-library.html?highlight=option#option). is there an Aptos-specific move guide?

Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:333]: Hehe, nope! I was able to find it in `aptos-core/aptos-move/framework/move-stdlib/sources/option.move`.

Be sure to look there for Move standard library features.

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/account.move:303]: exists_at(addr) not necessary?

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/account.move:587]: could we add assert to show bob has the capability?

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:303]: It might be possible there isn't an `Account` resource there though...

Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/account.move:303]: it isn't possible in our current framework to create a signer without an account.

but it isn't great to have the same error for two different errors

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/account.move:308]: well it isn't being used, so we could try to amend it and see if you can pass the `e2e-move-tests` for upgradeability
it isn't strictly necessary since the public key already ties the proof to the account but it is a good practice

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:330]: this feels like a lot of reusable boilerplate code... @alinush are you going to make a general library or should we just add that?

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/account.move:333]: yes, there is a great aptos-specific move guide... the code :trollface:

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/account.move:586]: Personally I don't think we should or need to validate that our crypto libraries are doing the right thing. I do agree that if we break something (which we aren't allowed to any more), then having gen libraries is great

**Review by davidiw — COMMENTED** [review]
I'm good, but you have two active reviewers, so I'll let them approve but poke me if need be :)

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:330]: Yes, I think we could abstract this either using this `account_scheme` parameter or with generics. Perhaps in a new `aptos_std::signatures` module?

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:586]: What do you mean by "gen libraries"?

**Review by davidiw — CHANGES_REQUESTED** [review]
PR hasn't been updated for a bit... we should get this out before first update

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:334]: just added the address of the source account - do we want to do the same for the `SignerCapabilityOfferProofChallenge` (which would be a breaking change)?

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:357]: @alinush are you going to make a gen lib or should we do that? if we do that, would you happen to have an example?

Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:373]: That will be a broader effort that should include domain separation too. It'll take more time...

**Comment by areshand** [comment]
I don't see any function using rotation capability to rotate the auth key. we should add these function after the capability is used somewhere

**Review by davidiw — COMMENTED** [review]
@chloeqjz / @alinush is the current plan to wait until we have the new domain separation?

**Comment by chloeqjz** [comment]
> @chloeqjz / @alinush is the current plan to wait until we have the new domain separation?

yes! we communicated on slack

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (lightmark) [inline aptos-move/framework/aptos-framework/sources/account.move:71]: Could you comment why `source_address` is necessary here?

File: aptos-move/e2e-move-tests/src/harness.rs
Comment (lightmark) [inline aptos-move/e2e-move-tests/src/harness.rs:114]: Why does this change relate to the PR.

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (lightmark) [inline aptos-move/framework/aptos-framework/sources/account.move:380]: why is `to_be_revoked_address` required? It seems simple w/o that.

File: aptos-move/framework/aptos-framework/sources/account.move
Comment (lightmark) [inline aptos-move/framework/aptos-framework/sources/account.move:405]: why delete `exists_at(source_address)`

**Pull request closed without merging**
